istio-interop-testing/src/main/java/io/grpc/testing/istio/EchoTestServer.java

```diff
@@ -0,0 +1,370 @@
+    } catch (IOException ex) {
+      logger.log(Level.INFO, "Failed to determine hostname. Will generate one", ex);
+    }
+    // let's make an identifier for ourselves.
+    return "generated-" + new Random().nextInt();
+  }
+
+  /**
+   * The main application allowing this program to be launched from the command line.
+   */
+  public static void main(String[] args) throws Exception {
+    Map<String, List<String>> processedArgs = preprocessArgs(args);
+    List<Integer> grpcPorts = getGrpcPorts(processedArgs);
+
+    String hostname = determineHostname();
+    EchoTestServer echoTestServer = new EchoTestServer();
+    echoTestServer.runServers(grpcPorts, hostname);
+    Runtime.getRuntime().addShutdownHook(new Thread(() -> {
+      try {
+        System.out.println("Shutting down");
+        echoTestServer.stopServers();
+      } catch (Exception e) {
+        logger.log(Level.SEVERE, "stopServers", e);
+        throw e;
+      }
+    }));
+    echoTestServer.blockUntilShutdown();
+  }
+
+  void runServers(List<Integer> grpcPorts, String hostname) throws IOException {
+    ServerServiceDefinition service = ServerInterceptors.intercept(
+        new EchoTestServiceImpl(hostname), new EchoTestServerInterceptor());
+    servers = new ArrayList<>(grpcPorts.size() + 1);
+    for (int port : grpcPorts) {
+      runServer(port, service);
+    }
+  }
+
+  void runServer(int port, ServerServiceDefinition service) throws IOException {
+    logger.log(Level.INFO, "Listening GRPC on " + port);
+    servers.add(Grpc.newServerBuilderForPort(port, InsecureServerCredentials.create())
+        .addService(service)
+        .build().start());
+  }
+
+  void stopServers() {
+    for (Server server : servers) {
+      server.shutdownNow();
+    }
+  }
+
+  void blockUntilShutdown() throws InterruptedException {
+    for (Server server : servers) {
+      if (!server.awaitTermination(5, TimeUnit.SECONDS)) {
+        System.err.println("Timed out waiting for server shutdown");
+      }
+    }
+  }
+
+  private static class EchoTestServerInterceptor implements ServerInterceptor {
+
+    @Override
+    public <ReqT, RespT> ServerCall.Listener<ReqT> interceptCall(ServerCall<ReqT, RespT> call,
+        final Metadata requestHeaders, ServerCallHandler<ReqT, RespT> next) {
+      final String methodName = call.getMethodDescriptor().getBareMethodName();
+
+      // we need this processing only for Echo
+      if (!"Echo".equals(methodName)) {
+        return next.startCall(call, requestHeaders);
+      }
+
+      final SocketAddress peerAddress = call.getAttributes()
+          .get(Grpc.TRANSPORT_ATTR_REMOTE_ADDR);
+
+      return next.startCall(
+          new ForwardingServerCall.SimpleForwardingServerCall<ReqT, RespT>(call) {
+
+            @SuppressWarnings("unchecked")
+            @Override
+            public void sendMessage(RespT message) {
+              io.grpc.testing.istio.Istio.EchoResponse echoResponse =
```

Comment:
done

Reply:
@ejona86  with this change a test fails on Linux (and probably Windows) but passes on MacOS in the CI. Also passes on my local build. Can you think of any reason why this change causes OS dependent behavior?

Reply:
It has nothing to do with "this change." As I messaged you earlier, the test was flaky when it was merged previously. For example, this run was on master: https://source.cloud.google.com/results/invocations/2f2fb97f-9783-4f73-9a59-25a3027fe22d

I don't expect this is OS-related. Instead it is probably a race.

Reply:
I think the problem is rpc timeout (of 100 ms) that's causing problems for this project with a single test class (so the JVM startup time is causing timeouts since nothing else runs before this)